hydra-node/src/Hydra/Chain/Direct/State.hs

```diff
@@ -131,6 +131,7 @@ class HasKnownUTxO a where
 data ChainStateAt = ChainStateAt
   { chainState :: ChainState
   , recordedAt :: Maybe ChainPoint
+  , previous :: Maybe ChainStateAt
```

Comment:
I would like to "make invalid state unrepresentable" here.
I think we should refactor ChainStateAt into a sum type to avoid it.

```hs
data ChainStateAt = 
  ChainStateAtZero
  | ChainStateAt
      { chainState :: ChainState
      , recordedAt :: ChainPoint
      , previous :: ChainStateAt
      }
```

Reply:
I like that mantra, but what you suggest is not the same. We can have a `ChainStateAt` which has a `ChainState` that was `recordedAt` some point.. but still has no `previous` entry.

```
data ChainStateAt 
  = ChainStateAtZero
    { chainState :: ChainState
    , recordedAt :: ChainPoint
    }
  | ChainStateAt
    { chainState :: ChainState
    , recordedAt :: ChainPoint
    , previous :: ChainStateAt
    }
```

Is an alternative, but that is not clearer than the `Maybe` in `previous IMO.